plugins/woocommerce/src/Blocks/BlockTypes/AddToCartWithOptionsVariationSelector.php

```diff
@@ -17,16 +17,348 @@ class AddToCartWithOptionsVariationSelector extends AbstractBlock {
+	}
+
+	/**
+	 * Get HTML for taxonomy-based options.
+	 *
+	 * @param WC_Product $product The product object.
+	 * @param string     $attribute_name Name of the attribute.
+	 * @param array      $options Available options.
+	 * @param string     $selected Selected value.
+	 * @return string Options HTML
+	 */
+	private function get_taxonomy_options_html( $product, $attribute_name, $options, $selected ): string {
+		$terms = wc_get_product_terms( $product->get_id(), $attribute_name, array( 'fields' => 'all' ) );
+		$html  = '';
+
+		foreach ( $terms as $term ) {
+			if ( in_array( $term->slug, $options, true ) ) {
+				$html .= sprintf(
+					'<option value="%s" %s>%s</option>',
+					esc_attr( $term->slug ),
+					selected( sanitize_title( $selected ), $term->slug, false ),
+					/**
+					 * Filter the variation option name.
+					 *
+					 * @since 9.7.0
+					 *
+					 * @param string   $term_name      The term name.
+					 * @param WP_Term  $term           Term object.
+					 * @param string   $attribute_name Name of the attribute.
+					 * @param WC_Product $product      Product object.
+					 */
+					esc_html( apply_filters( 'woocommerce_variation_option_name', $term->name, $term, $attribute_name, $product ) )
+				);
+			}
+		}
+
+		return $html;
+	}
+
+	/**
+	 * Get HTML for custom attribute options.
+	 *
+	 * @param WC_Product $product The product object.
+	 * @param string     $attribute_name Name of the attribute.
+	 * @param array      $options Available options.
+	 * @param string     $selected Selected value.
+	 * @return string Options HTML
+	 */
+	private function get_custom_options_html( $product, $attribute_name, $options, $selected ): string {
+		$html = '';
+
+		foreach ( $options as $option ) {
+			$selected_value = sanitize_title( $selected ) === $selected
+				? selected( $selected, sanitize_title( $option ), false )
+				: selected( $selected, $option, false );
+
+			$html .= sprintf(
+				'<option value="%s" %s>%s</option>',
+				esc_attr( $option ),
+				$selected_value,
+				/**
+				 * Filter the variation option name.
+				 *
+				 * @since 9.7.0
+				 *
+				 * @param string   $option_name      The option name.
+				 * @param string   $option           The option value.
+				 * @param string   $attribute_name Name of the attribute.
+				 * @param WC_Product $product      Product object.
+				 */
+				esc_html( apply_filters( 'woocommerce_variation_option_name', $option, null, $attribute_name, $product ) )
+			);
+		}
+
+		return $html;
+	}
+
+	/**
+	 * Render variation form.
+	 *
+	 * @param WC_Product $product Product instance.
+	 * @param array      $attributes Block attributes.
+	 * @return string Rendered form HTML
+	 */
+	private function render_variation_form( $product, $attributes ): string {
+		$variation_attributes = $product->get_variation_attributes();
+		if ( empty( $variation_attributes ) ) {
+			return '';
+		}
+
+		wp_enqueue_script( 'wc-add-to-cart-variation' );
```

Comment:
We should rewrite [this file](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/legacy/js/frontend/add-to-cart-variation.js). I'll create an issue to tackle that.